zookeeper-server/src/main/java/org/apache/zookeeper/ZKUtil.java

```diff
@@ -45,20 +48,67 @@
      *
      * @throws IllegalArgumentException if an invalid path is specified
      */
-    public static void deleteRecursive(ZooKeeper zk, final String pathRoot)
+    public static boolean deleteRecursive(ZooKeeper zk, final String pathRoot)
         throws InterruptedException, KeeperException
     {
         PathUtils.validatePath(pathRoot);
 
         List<String> tree = listSubTreeBFS(zk, pathRoot);
         LOG.debug("Deleting " + tree);
         LOG.debug("Deleting " + tree.size() + " subnodes ");
-        for (int i = tree.size() - 1; i >= 0 ; --i) {
-            //Delete the leaves first and eventually get rid of the root
-            zk.delete(tree.get(i), -1); //Delete all versions of the node with -1.
+
+        int asyncReqRateLimit = 10;
+        // Try deleting the tree nodes in batches of size 1000.
+        // If some batch failed, try again with batches of size 1 to delete as
+        // many nodes as possible.
```

Comment:
To help capture a plausible scenario, I've added a new unit test as you suggest.

Using a multi-op batch mode is bound to be more brittle than using single operations. As a user, I'd rather that the tool retry in the more resilient mode if the less resilient suffered a failure rather than be asked to debug/retry manually.

Reply:
I agree with @szepet : I don't think we need this retry logic in the cli.
I would just let the user know about the command failed because of this and that and let him decide whether he/she wants to retry with a different batch size.

Reply:
Oh, this is not the cli, but `ZKUtil`, so essentially it's a public api. Another reason not to be proactive.

Reply:
Now that batch mode is configurable, this more conservative approach is making more sense to me. 

Alternate idea - the tool swallows all errors until it has tried deleting every node and then presents all errors as a single aggregate instead of halting at the first error and presenting it.